// @vue/devtools-api/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
let _vue_devtools_kit = require("@vue/devtools-kit");
Object.defineProperty(exports, "addCustomCommand", {
	enumerable: true,
	get: function() {
		return _vue_devtools_kit.addCustomCommand;
	}
});
Object.defineProperty(exports, "addCustomTab", {
	enumerable: true,
	get: function() {
		return _vue_devtools_kit.addCustomTab;
	}
});
Object.defineProperty(exports, "onDevToolsClientConnected", {
	enumerable: true,
	get: function() {
		return _vue_devtools_kit.onDevToolsClientConnected;
	}
});
Object.defineProperty(exports, "onDevToolsConnected", {
	enumerable: true,
	get: function() {
		return _vue_devtools_kit.onDevToolsConnected;
	}
});
Object.defineProperty(exports, "removeCustomCommand", {
	enumerable: true,
	get: function() {
		return _vue_devtools_kit.removeCustomCommand;
	}
});
Object.defineProperty(exports, "setupDevToolsPlugin", {
	enumerable: true,
	get: function() {
		return _vue_devtools_kit.setupDevToolsPlugin;
	}
});
Object.defineProperty(exports, "setupDevtoolsPlugin", {
	enumerable: true,
	get: function() {
		return _vue_devtools_kit.setupDevToolsPlugin;
	}
});
