// @nextcloud/vue/dist/Composables/useFormatDateTime.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const l10n = require("@nextcloud/l10n");
const core = require("@vueuse/core");
const Vue = require("vue");
const _l10n = require("../chunks/_l10n-skrZri3h.cjs");
_l10n.register(_l10n.t2);
const FEW_SECONDS_AGO = {
  long: _l10n.t("a few seconds ago"),
  short: _l10n.t("seconds ago"),
  // FOR TRANSLATORS: Shorter version of 'a few seconds ago'
  narrow: _l10n.t("sec. ago")
  // FOR TRANSLATORS: If possible in your language an even shorter version of 'a few seconds ago'
};
function useFormatRelativeTime(timestamp = Date.now(), opts = {}) {
  let timeoutId;
  const date = Vue.computed(() => new Date(core.toValue(timestamp)));
  const options = Vue.computed(() => {
    const { language, relativeTime: relativeTime2, ignoreSeconds } = core.toValue(opts);
    return {
      ...language && { language },
      ...relativeTime2 && { relativeTime: relativeTime2 },
      ignoreSeconds: ignoreSeconds ? FEW_SECONDS_AGO[relativeTime2 || "long"] : false
    };
  });
  const relativeTime = Vue.ref("");
  Vue.watchEffect(() => updateRelativeTime());
  function updateRelativeTime() {
    relativeTime.value = l10n.formatRelativeTime(date.value, options.value);
    if (core.toValue(opts).update !== false) {
      const diff = Math.abs(Date.now() - new Date(core.toValue(timestamp)).getTime());
      const interval = diff > 12e4 || options.value.ignoreSeconds ? Math.min(diff / 60, 18e5) : 1e3;
      timeoutId = window.setTimeout(updateRelativeTime, interval);
    }
  }
  Vue.onUnmounted(() => timeoutId && window.clearTimeout(timeoutId));
  return Vue.readonly(relativeTime);
}
function useFormatTime(timestamp, opts) {
  const options = Vue.computed(() => ({
    locale: l10n.getCanonicalLocale(),
    format: { dateStyle: "short", timeStyle: "medium" },
    ...core.toValue(opts)
  }));
  const formatter = Vue.computed(() => new Intl.DateTimeFormat(options.value.locale, options.value.format));
  return Vue.computed(() => formatter.value.format(core.toValue(timestamp)));
}
function useFormatDateTime(timestamp = Date.now(), opts = {}) {
  const formattedFullTime = useFormatTime(timestamp, opts);
  const relativeTime = useFormatRelativeTime(timestamp, Vue.computed(() => {
    const options = core.toValue(opts);
    return {
      ...options,
      relativeTime: typeof options.relativeTime === "string" ? options.relativeTime : "long"
    };
  }));
  const formattedTime = Vue.computed(() => core.toValue(opts).relativeTime !== false ? relativeTime.value : formattedFullTime.value);
  return {
    formattedTime,
    formattedFullTime
  };
}
exports.useFormatDateTime = useFormatDateTime;
exports.useFormatRelativeTime = useFormatRelativeTime;
exports.useFormatTime = useFormatTime;
