// @nextcloud/vue/dist/chunks/NcAppSettingsSection-CDLQ-AjK.cjs
require('../assets/NcAppSettingsSection-Yg4NYCih.css');
"use strict";
const Vue = require("vue");
const logger = require("./logger-3HuiEIF6.cjs");
const useAppSettingsDialog = require("./useAppSettingsDialog-DKwz-cwX.cjs");
const _pluginVue2_normalizer = require("./_plugin-vue2_normalizer-V0q-tHlQ.cjs");
const _sfc_main = /* @__PURE__ */ Vue.defineComponent({
  __name: "NcAppSettingsSection",
  props: {
    name: null,
    description: null,
    id: null,
    order: null
  },
  setup(__props) {
    const props = __props;
    const slots = Vue.useSlots();
    const { registerSection, unregisterSection } = useAppSettingsDialog.useAppSettingsDialog();
    const legacy = Vue.inject(useAppSettingsDialog.APP_SETTINGS_LEGACY_DESIGN_KEY);
    const htmlId = Vue.computed(() => "settings-section_" + props.id);
    Vue.watch(() => props.id, () => {
      if (!/^[a-z0-9\-_]+$/.test(props.id)) {
        logger.logger.warn(`Invalid id prop: ${props.id}. Only alphanumeric, dash and underscore are allowed.`);
      }
    }, { immediate: true });
    Vue.watch([() => props.id, () => props.name, () => props.order], ([newId, newName, newOrder], [oldId, ,]) => {
      unregisterSection(oldId);
      registerSection(newId, newName, newOrder, slots?.icon?.());
    });
    Vue.onMounted(() => {
      registerSection(props.id, props.name, props.order, slots?.icon?.());
    });
    Vue.onBeforeUnmount(() => {
      unregisterSection(props.id);
    });
    return { __sfc: true, props, slots, registerSection, unregisterSection, legacy, htmlId };
  }
});
var _sfc_render = function render() {
  var _vm = this, _c = _vm._self._c, _setup = _vm._self._setupProxy;
  return _c("section", { staticClass: "app-settings-section", class: { "app-settings-section__legacy": _setup.legacy }, attrs: { "id": _setup.htmlId, "aria-labelledby": `${_setup.htmlId}--label` } }, [_c("h3", { staticClass: "app-settings-section__name", attrs: { "id": `${_setup.htmlId}--label` } }, [_vm._v(" " + _vm._s(_vm.name) + " ")]), _c("div", { staticClass: "app-settings-section__description" }, [_vm._v(" " + _vm._s(_vm.description) + " ")]), _c("div", { staticClass: "app-settings-section__content" }, [_vm._t("default")], 2), _vm._e()], 2);
};
var _sfc_staticRenderFns = [];
var __component__ = /* @__PURE__ */ _pluginVue2_normalizer.normalizeComponent(
  _sfc_main,
  _sfc_render,
  _sfc_staticRenderFns,
  false,
  null,
  "06cd755d"
);
const NcAppSettingsSection = __component__.exports;
exports.NcAppSettingsSection = NcAppSettingsSection;
