// @nextcloud/vue/dist/chunks/NcAppNavigationSettings-CotFe5aD.cjs
require('../assets/NcMentionBubble-CaztX9Pv.css');
require('../assets/NcAppNavigationSettings-2Wh1E3Hq.css');
"use strict";
const mdi = require("./mdi-DTiP6a6e.cjs");
const components = require("@vueuse/components");
const NcButton = require("./NcButton-ByTFUhwJ.cjs");
const NcIconSvgWrapper = require("./NcIconSvgWrapper-Bbl4kXZI.cjs");
const _l10n = require("./_l10n-skrZri3h.cjs");
const Mixins_clickOutsideOptions = require("../Mixins/clickOutsideOptions.cjs");
require("../Composables/useIsFullscreen.cjs");
require("../Composables/useIsMobile.cjs");
require("escape-html");
require("striptags");
require("vue");
require("../Composables/useIsDarkTheme.cjs");
require("@nextcloud/router");
require("../Functions/isDarkTheme.cjs");
;/* empty css                                                                           */
const _pluginVue2_normalizer = require("./_plugin-vue2_normalizer-V0q-tHlQ.cjs");
require("@nextcloud/auth");
require("@nextcloud/axios");
require("@nextcloud/capabilities");
require("./logger-3HuiEIF6.cjs");
const GenRandomId = require("./GenRandomId-D7iOvpZS.cjs");
const legacy = require("./legacy-Bk0qQxIg.cjs");
_l10n.register(_l10n.t48);
const _sfc_main = {
  directives: {
    ClickOutside: components.vOnClickOutside
  },
  components: {
    NcButton: NcButton.NcButton,
    NcIconSvgWrapper: NcIconSvgWrapper.NcIconSvgWrapper
  },
  mixins: [
    Mixins_clickOutsideOptions
  ],
  props: {
    /**
     * Text of the button
     *
     * @default 'Settings'
     */
    name: {
      type: String,
      required: false,
      default: _l10n.t("Settings")
    }
  },
  setup() {
    const contentId = GenRandomId.GenRandomId();
    return {
      contentId,
      isLegacy32: legacy.isLegacy32,
      mdiCog: mdi.mdiCog,
      mdiCogOutline: mdi.mdiCogOutline
    };
  },
  data() {
    return {
      open: false
    };
  },
  computed: {
    clickOutsideConfig() {
      return [
        this.closeMenu,
        this.clickOutsideOptions
      ];
    }
  },
  methods: {
    closeMenu() {
      this.open = false;
    }
  }
};
var _sfc_render = function render() {
  var _vm = this, _c = _vm._self._c;
  return _c("div", { directives: [{ name: "click-outside", rawName: "v-click-outside", value: _vm.clickOutsideConfig, expression: "clickOutsideConfig" }], class: { open: _vm.open }, attrs: { "id": "app-settings" } }, [_c("div", { attrs: { "id": "app-settings__header" } }, [_c("NcButton", { staticClass: "settings-button", attrs: { "aria-controls": _vm.contentId, "aria-expanded": _vm.open ? "true" : "false", "alignment": "start", "variant": "tertiary", "wide": "" }, on: { "click": function($event) {
    _vm.open = !_vm.open;
  } }, scopedSlots: _vm._u([{ key: "icon", fn: function() {
    return [_c("NcIconSvgWrapper", { staticClass: "settings-button__icon", attrs: { "path": _vm.isLegacy32 ? _vm.mdiCog : _vm.mdiCogOutline } })];
  }, proxy: true }]) }, [_vm._v(" " + _vm._s(_vm.name) + " ")])], 1), _c("Transition", { attrs: { "name": "slide-up" } }, [_c("div", { directives: [{ name: "show", rawName: "v-show", value: _vm.open, expression: "open" }], attrs: { "id": _vm.contentId } }, [_vm._t("default")], 2)])], 1);
};
var _sfc_staticRenderFns = [];
var __component__ = /* @__PURE__ */ _pluginVue2_normalizer.normalizeComponent(
  _sfc_main,
  _sfc_render,
  _sfc_staticRenderFns,
  false,
  null,
  "8fb21c8b"
);
const NcAppNavigationSettings = __component__.exports;
exports.NcAppNavigationSettings = NcAppNavigationSettings;
