// @nextcloud/vue/dist/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const NcActionButton = require("./chunks/NcActionButton-CiYel4_C.cjs");
const NcActionButtonGroup = require("./chunks/NcActionButtonGroup-CFGCmzTM.cjs");
const Components_NcActionCaption = require("./Components/NcActionCaption.cjs");
const NcActionCheckbox = require("./chunks/NcActionCheckbox-BGCUJhg_.cjs");
const NcActionInput = require("./chunks/NcActionInput-CcA3XFGe.cjs");
const Components_NcActionLink = require("./Components/NcActionLink.cjs");
const NcActionRadio = require("./chunks/NcActionRadio-DbkRiTHX.cjs");
const Components_NcActionRouter = require("./Components/NcActionRouter.cjs");
const NcActions = require("./chunks/NcActions-D51W1X9v.cjs");
const Components_NcActionSeparator = require("./Components/NcActionSeparator.cjs");
const Components_NcActionText = require("./Components/NcActionText.cjs");
const NcActionTextEditable = require("./chunks/NcActionTextEditable-BqJ5wONY.cjs");
const NcAppContent = require("./chunks/NcAppContent-0hbQMH35.cjs");
const Components_NcAppContentDetails = require("./Components/NcAppContentDetails.cjs");
const Components_NcAppContentList = require("./Components/NcAppContentList.cjs");
const NcAppNavigation = require("./chunks/NcAppNavigation-DB3HqxPL.cjs");
const NcAppNavigationCaption = require("./chunks/NcAppNavigationCaption-WVnxSeIf.cjs");
const Components_NcAppNavigationIconBullet = require("./Components/NcAppNavigationIconBullet.cjs");
const NcAppNavigationItem = require("./chunks/NcAppNavigationItem-0zDqIuUZ.cjs");
const Components_NcAppNavigationList = require("./Components/NcAppNavigationList.cjs");
const NcAppNavigationNew = require("./chunks/NcAppNavigationNew-DRdPwKOi.cjs");
const NcAppNavigationNewItem = require("./chunks/NcAppNavigationNewItem-SR744REn.cjs");
const NcAppNavigationSearch = require("./chunks/NcAppNavigationSearch-Czcwiin4.cjs");
const NcAppNavigationSettings = require("./chunks/NcAppNavigationSettings-CotFe5aD.cjs");
const Components_NcAppNavigationSpacer = require("./Components/NcAppNavigationSpacer.cjs");
const NcAppSettingsDialog = require("./chunks/NcAppSettingsDialog-Fa5DTfig.cjs");
const NcAppSettingsSection = require("./chunks/NcAppSettingsSection-CDLQ-AjK.cjs");
const NcAppSettingsShortcutsSection = require("./chunks/NcAppSettingsShortcutsSection-DgGU8-cF.cjs");
const NcAppSidebar = require("./chunks/NcAppSidebar-DXEl0_ou.cjs");
const NcAppSidebarHeader = require("./chunks/NcAppSidebarHeader-D-c_yxvx.cjs");
const Components_NcAppSidebarTab = require("./Components/NcAppSidebarTab.cjs");
const NcAssistantButton = require("./chunks/NcAssistantButton-XS3UdAKm.cjs");
const Components_NcAssistantContent = require("./Components/NcAssistantContent.cjs");
const NcAssistantIcon = require("./chunks/NcAssistantIcon-CD5mu9O1.cjs");
const NcAvatar = require("./chunks/NcAvatar-D09ZAgeJ.cjs");
const NcBlurHash = require("./chunks/NcBlurHash-D0wcwBet.cjs");
const NcBreadcrumb = require("./chunks/NcBreadcrumb-14OmaYrH.cjs");
const NcBreadcrumbs = require("./chunks/NcBreadcrumbs-CZt3UuyO.cjs");
const NcButton = require("./chunks/NcButton-ByTFUhwJ.cjs");
const NcCheckboxRadioSwitch = require("./chunks/NcCheckboxRadioSwitch-BQeglWXd.cjs");
const NcCollectionList = require("./chunks/NcCollectionList-2jk5XCm8.cjs");
const NcColorPicker = require("./chunks/NcColorPicker-BTiAa5mm.cjs");
const NcContent = require("./chunks/NcContent-BB-v-amq.cjs");
const NcCounterBubble = require("./chunks/NcCounterBubble-hyvazclA.cjs");
const NcDashboardWidget = require("./chunks/NcDashboardWidget-dMwnFBzc.cjs");
const NcDashboardWidgetItem = require("./chunks/NcDashboardWidgetItem-9tQgaPXS.cjs");
const NcDateTime = require("./chunks/NcDateTime-DE-ltgw2.cjs");
const Components_NcDateTimePicker = require("./Components/NcDateTimePicker.cjs");
const Components_NcDateTimePickerNative = require("./Components/NcDateTimePickerNative.cjs");
const NcDialog = require("./chunks/NcDialog-Qak1QTrv.cjs");
const NcDialogButton = require("./chunks/NcDialogButton-BoetK2zm.cjs");
const Components_NcEllipsisedOption = require("./Components/NcEllipsisedOption.cjs");
const NcEmojiPicker = require("./chunks/NcEmojiPicker-2KIJVstd.cjs");
const Components_NcEmptyContent = require("./Components/NcEmptyContent.cjs");
const NcFilePicker = require("./chunks/NcFilePicker-BL6iajt2.cjs");
const NcFormBox = require("./chunks/NcFormBox-DRGVpFeZ.cjs");
const NcFormBoxButton = require("./chunks/NcFormBoxButton-ozVuTgCZ.cjs");
const NcFormBoxCopyButton = require("./chunks/NcFormBoxCopyButton-BS9aM994.cjs");
const NcFormBoxSwitch = require("./chunks/NcFormBoxSwitch-7R8iAj1R.cjs");
const NcFormGroup = require("./chunks/NcFormGroup-AxkxubKD.cjs");
const Components_NcGuestContent = require("./Components/NcGuestContent.cjs");
const NcHeaderButton = require("./chunks/NcHeaderButton-BndVfb3e.cjs");
const NcHeaderMenu = require("./chunks/NcHeaderMenu-DUQ4pqU-.cjs");
const Components_NcHighlight = require("./chunks/index-Dr9k4hIP.cjs");
const NcHotkey = require("./chunks/NcHotkey-vKyDZ2a7.cjs");
const NcHotkeyList = require("./chunks/NcHotkeyList-DhFQdopw.cjs");
const NcIconSvgWrapper = require("./chunks/NcIconSvgWrapper-Bbl4kXZI.cjs");
const NcInputField = require("./chunks/NcInputField-Brpn446B.cjs");
const NcListItem = require("./chunks/NcListItem-C-e9iA9w.cjs");
const NcListItemIcon = require("./chunks/NcListItemIcon-BQ0f65Vg.cjs");
const Components_NcLoadingIcon = require("./Components/NcLoadingIcon.cjs");
const Components_NcModal = require("./Components/NcModal.cjs");
const Components_NcNoteCard = require("./Components/NcNoteCard.cjs");
const NcPasswordField = require("./chunks/NcPasswordField-PQt7wMbW.cjs");
const NcPopover = require("./chunks/NcPopover-Ujf9nIvN.cjs");
const NcProgressBar = require("./chunks/NcProgressBar-DvxA9bq5.cjs");
const NcRadioGroup = require("./chunks/NcRadioGroup-CsPmwGlc.cjs");
const NcRadioGroupButton = require("./chunks/NcRadioGroupButton-6skMHhr7.cjs");
const NcRelatedResourcesPanel = require("./chunks/NcRelatedResourcesPanel-l1AlGr83.cjs");
const Mixins_richEditor = require("./chunks/index-phHwkniW.cjs");
const NcRichContenteditable = require("./chunks/NcRichContenteditable-DZjIiOKo.cjs");
require("@nextcloud/auth");
require("@nextcloud/axios");
require("@nextcloud/router");
require("@nextcloud/sharing/public");
require("@vueuse/core");
require("vue");
require("vue-router");
const referencePickerModal = require("./chunks/referencePickerModal-BGYbZYrX.cjs");
const customPickerElements = require("./chunks/customPickerElements-GyIIOHiQ.cjs");
require("unist-builder");
require("unist-util-visit-parents");
require("./chunks/logger-3HuiEIF6.cjs");
const NcRichText = require("./chunks/NcRichText-pqvg5l27.cjs");
const NcSelect = require("./chunks/NcSelect-BCcSawYf.cjs");
const NcTextField = require("./chunks/NcTextField-B92lM77I.cjs");
require("@nextcloud/event-bus");
const NcSavingIndicatorIcon = require("./chunks/NcSavingIndicatorIcon-DFyB2KRK.cjs");
const NcSelectTags = require("./chunks/NcSelectTags-uU_mHbSm.cjs");
const NcSelectUsers = require("./chunks/NcSelectUsers-BB9ZIHeB.cjs");
const NcSettingsInputText = require("./chunks/NcSettingsInputText-BF5437Q7.cjs");
const NcSettingsSection = require("./chunks/NcSettingsSection-BrIKMh-B.cjs");
const NcSettingsSelectGroup = require("./chunks/NcSettingsSelectGroup-BtNFBX1V.cjs");
const NcTextArea = require("./chunks/NcTextArea-B9Vn3rQC.cjs");
const NcThemeProvider = require("./chunks/NcThemeProvider-BOKkNukO.cjs");
const NcTimezonePicker = require("./chunks/NcTimezonePicker-D2e7hW-Z.cjs");
const NcUserBubble = require("./chunks/NcUserBubble-CNQJUk2l.cjs");
const NcUserStatusIcon = require("./chunks/NcUserStatusIcon-BnGtoiCy.cjs");
const Components_NcVNodes = require("./Components/NcVNodes.cjs");
const Composables_useFormatDateTime = require("./Composables/useFormatDateTime.cjs");
const Composables_useHotKey = require("./Composables/useHotKey.cjs");
const Composables_useIsDarkTheme = require("./Composables/useIsDarkTheme.cjs");
const Composables_useIsFullscreen = require("./Composables/useIsFullscreen.cjs");
const Composables_useIsMobile = require("./Composables/useIsMobile.cjs");
const Functions_a11y = require("./Functions/a11y.cjs");
const Functions_contactsMenu = require("./Functions/contactsMenu.cjs");
const Functions_dialog = require("./Functions/dialog.cjs");
const emoji = require("./chunks/emoji-aWWWqkrJ.cjs");
const Functions_isDarkTheme = require("./Functions/isDarkTheme.cjs");
const Functions_preloadImage = require("./Functions/preloadImage.cjs");
const Functions_usernameToColor = require("./Functions/usernameToColor.cjs");
const Directives_Focus = require("./Directives/Focus.cjs");
const Directives_Linkify = require("./Directives/Linkify.cjs");
require("./Directives/Tooltip.cjs");
const Mixins_clickOutsideOptions = require("./Mixins/clickOutsideOptions.cjs");
const Mixins_isFullscreen = require("./Mixins/isFullscreen.cjs");
const Mixins_isMobile = require("./Mixins/isMobile.cjs");
const floatingVue = require("floating-vue");
const NcComponents = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  NcActionButton: NcActionButton.NcActionButton,
  NcActionButtonGroup: NcActionButtonGroup.NcActionButtonGroup,
  NcActionCaption: Components_NcActionCaption,
  NcActionCheckbox: NcActionCheckbox.NcActionCheckbox,
  NcActionInput: NcActionInput.NcActionInput,
  NcActionLink: Components_NcActionLink,
  NcActionRadio: NcActionRadio.NcActionRadio,
  NcActionRouter: Components_NcActionRouter,
  NcActionSeparator: Components_NcActionSeparator,
  NcActionText: Components_NcActionText,
  NcActionTextEditable: NcActionTextEditable.NcActionTextEditable,
  NcActions: NcActions.NcActions,
  NcAppContent: NcAppContent.NcAppContent,
  NcAppContentDetails: Components_NcAppContentDetails,
  NcAppContentList: Components_NcAppContentList,
  NcAppNavigation: NcAppNavigation.NcAppNavigation,
  NcAppNavigationCaption: NcAppNavigationCaption.NcAppNavigationCaption,
  NcAppNavigationIconBullet: Components_NcAppNavigationIconBullet,
  NcAppNavigationItem: NcAppNavigationItem.NcAppNavigationItem,
  NcAppNavigationList: Components_NcAppNavigationList,
  NcAppNavigationNew: NcAppNavigationNew.NcAppNavigationNew,
  NcAppNavigationNewItem: NcAppNavigationNewItem.NcAppNavigationNewItem,
  NcAppNavigationSearch: NcAppNavigationSearch.NcAppNavigationSearch,
  NcAppNavigationSettings: NcAppNavigationSettings.NcAppNavigationSettings,
  NcAppNavigationSpacer: Components_NcAppNavigationSpacer,
  NcAppSettingsDialog: NcAppSettingsDialog.NcAppSettingsDialog,
  NcAppSettingsSection: NcAppSettingsSection.NcAppSettingsSection,
  NcAppSettingsSectionShortcuts: NcAppSettingsShortcutsSection.NcAppSettingsShortcutsSection,
  NcAppSettingsShortcutsSection: NcAppSettingsShortcutsSection.NcAppSettingsShortcutsSection,
  NcAppSidebar: NcAppSidebar.NcAppSidebar,
  NcAppSidebarHeader: NcAppSidebarHeader.NcAppSidebarHeader,
  NcAppSidebarTab: Components_NcAppSidebarTab,
  NcAssistantButton: NcAssistantButton.NcAssistantButton,
  NcAssistantContent: Components_NcAssistantContent,
  NcAssistantIcon: NcAssistantIcon.NcAssistantIcon,
  NcAutoCompleteResult: NcRichContenteditable.NcAutoCompleteResult,
  NcAvatar: NcAvatar.NcAvatar,
  NcBlurHash: NcBlurHash.NcBlurHash,
  NcBreadcrumb: NcBreadcrumb.NcBreadcrumb,
  NcBreadcrumbs: NcBreadcrumbs.NcBreadcrumbs,
  NcButton: NcButton.NcButton,
  NcCheckboxRadioSwitch: NcCheckboxRadioSwitch.NcCheckboxRadioSwitch,
  NcCollectionList: NcCollectionList.NcCollectionList,
  NcColorPicker: NcColorPicker.NcColorPicker,
  NcContent: NcContent.NcContent,
  NcCounterBubble: NcCounterBubble.NcCounterBubble,
  NcDashboardWidget: NcDashboardWidget.NcDashboardWidget,
  NcDashboardWidgetItem: NcDashboardWidgetItem.NcDashboardWidgetItem,
  NcDateTime: NcDateTime.NcDateTime,
  NcDateTimePicker: Components_NcDateTimePicker,
  NcDateTimePickerNative: Components_NcDateTimePickerNative,
  NcDialog: NcDialog.NcDialog,
  NcDialogButton: NcDialogButton.NcDialogButton,
  NcEllipsisedOption: Components_NcEllipsisedOption,
  NcEmojiPicker: NcEmojiPicker.NcEmojiPicker,
  NcEmptyContent: Components_NcEmptyContent,
  NcFilePicker: NcFilePicker.NcFilePicker,
  NcFormBox: NcFormBox.NcFormBox,
  NcFormBoxButton: NcFormBoxButton.NcFormBoxButton,
  NcFormBoxCopyButton: NcFormBoxCopyButton.NcFormBoxCopyButton,
  NcFormBoxSwitch: NcFormBoxSwitch.NcFormBoxSwitch,
  NcFormGroup: NcFormGroup.NcFormGroup,
  NcGuestContent: Components_NcGuestContent,
  NcHeaderButton: NcHeaderButton.NcHeaderButton,
  NcHeaderMenu: NcHeaderMenu.NcHeaderMenu,
  NcHighlight: Components_NcHighlight.NcHighlight,
  NcHotkey: NcHotkey.NcHotkey,
  NcHotkeyList: NcHotkeyList.NcHotkeyList,
  NcIconSvgWrapper: NcIconSvgWrapper.NcIconSvgWrapper,
  NcInputField: NcInputField.NcInputField,
  NcListItem: NcListItem.NcListItem,
  NcListItemIcon: NcListItemIcon.NcListItemIcon,
  NcLoadingIcon: Components_NcLoadingIcon,
  NcMentionBubble: Mixins_richEditor.NcMentionBubble,
  NcModal: Components_NcModal,
  NcNoteCard: Components_NcNoteCard,
  NcPasswordField: NcPasswordField.NcPasswordField,
  NcPopover: NcPopover.NcPopover,
  NcProgressBar: NcProgressBar.NcProgressBar,
  NcRadioGroup: NcRadioGroup.NcRadioGroup,
  NcRadioGroupButton: NcRadioGroupButton.NcRadioGroupButton,
  NcRelatedResourcesPanel: NcRelatedResourcesPanel.NcRelatedResourcesPanel,
  NcRichContenteditable: NcRichContenteditable.NcRichContenteditable,
  NcRichText: NcRichText.NcRichText,
  NcSavingIndicatorIcon: NcSavingIndicatorIcon.NcSavingIndicatorIcon,
  NcSelect: NcSelect.NcSelect,
  NcSelectTags: NcSelectTags.NcSelectTags,
  NcSelectUsers: NcSelectUsers.NcSelectUsers,
  NcSettingsInputText: NcSettingsInputText.NcSettingsInputText,
  NcSettingsSection: NcSettingsSection.NcSettingsSection,
  NcSettingsSelectGroup: NcSettingsSelectGroup.NcSettingsSelectGroup,
  NcTextArea: NcTextArea.NcTextArea,
  NcTextField: NcTextField.NcTextField,
  NcThemeProvider: NcThemeProvider.NcThemeProvider,
  NcTimezonePicker: NcTimezonePicker.NcTimezonePicker,
  NcUserBubble: NcUserBubble.NcUserBubble,
  NcUserStatusIcon: NcUserStatusIcon.NcUserStatusIcon,
  NcVNodes: Components_NcVNodes
}, Symbol.toStringTag, { value: "Module" }));
const NcDirectives = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  Focus: Directives_Focus,
  Linkify: Directives_Linkify,
  Tooltip: floatingVue.VTooltip
}, Symbol.toStringTag, { value: "Module" }));
const NextcloudVuePlugin = {
  install(Vue) {
    Object.entries(NcComponents).forEach(([name, component]) => {
      Vue.component(component.name || name, component);
    });
    Object.entries(NcDirectives).forEach(([name, directive]) => {
      Vue.directive(name, directive);
    });
  }
};
exports.NcActionButton = NcActionButton.NcActionButton;
exports.NcActionButtonGroup = NcActionButtonGroup.NcActionButtonGroup;
exports.NcActionCaption = Components_NcActionCaption;
exports.NcActionCheckbox = NcActionCheckbox.NcActionCheckbox;
exports.NcActionInput = NcActionInput.NcActionInput;
exports.NcActionLink = Components_NcActionLink;
exports.NcActionRadio = NcActionRadio.NcActionRadio;
exports.NcActionRouter = Components_NcActionRouter;
exports.NcActions = NcActions.NcActions;
exports.NcActionSeparator = Components_NcActionSeparator;
exports.NcActionText = Components_NcActionText;
exports.NcActionTextEditable = NcActionTextEditable.NcActionTextEditable;
exports.NcAppContent = NcAppContent.NcAppContent;
exports.NcAppContentDetails = Components_NcAppContentDetails;
exports.NcAppContentList = Components_NcAppContentList;
exports.NcAppNavigation = NcAppNavigation.NcAppNavigation;
exports.NcAppNavigationCaption = NcAppNavigationCaption.NcAppNavigationCaption;
exports.NcAppNavigationIconBullet = Components_NcAppNavigationIconBullet;
exports.NcAppNavigationItem = NcAppNavigationItem.NcAppNavigationItem;
exports.NcAppNavigationList = Components_NcAppNavigationList;
exports.NcAppNavigationNew = NcAppNavigationNew.NcAppNavigationNew;
exports.NcAppNavigationNewItem = NcAppNavigationNewItem.NcAppNavigationNewItem;
exports.NcAppNavigationSearch = NcAppNavigationSearch.NcAppNavigationSearch;
exports.NcAppNavigationSettings = NcAppNavigationSettings.NcAppNavigationSettings;
exports.NcAppNavigationSpacer = Components_NcAppNavigationSpacer;
exports.NcAppSettingsDialog = NcAppSettingsDialog.NcAppSettingsDialog;
exports.NcAppSettingsSection = NcAppSettingsSection.NcAppSettingsSection;
exports.NcAppSettingsSectionShortcuts = NcAppSettingsShortcutsSection.NcAppSettingsShortcutsSection;
exports.NcAppSettingsShortcutsSection = NcAppSettingsShortcutsSection.NcAppSettingsShortcutsSection;
exports.NcAppSidebar = NcAppSidebar.NcAppSidebar;
exports.NcAppSidebarHeader = NcAppSidebarHeader.NcAppSidebarHeader;
exports.NcAppSidebarTab = Components_NcAppSidebarTab;
exports.NcAssistantButton = NcAssistantButton.NcAssistantButton;
exports.NcAssistantContent = Components_NcAssistantContent;
exports.NcAssistantIcon = NcAssistantIcon.NcAssistantIcon;
exports.NcAvatar = NcAvatar.NcAvatar;
exports.userStatus = NcAvatar.userStatus;
exports.NcBlurHash = NcBlurHash.NcBlurHash;
exports.NcBreadcrumb = NcBreadcrumb.NcBreadcrumb;
exports.NcBreadcrumbs = NcBreadcrumbs.NcBreadcrumbs;
exports.NcButton = NcButton.NcButton;
exports.NcCheckboxRadioSwitch = NcCheckboxRadioSwitch.NcCheckboxRadioSwitch;
exports.NcCollectionList = NcCollectionList.NcCollectionList;
exports.NcColorPicker = NcColorPicker.NcColorPicker;
exports.NcContent = NcContent.NcContent;
exports.NcCounterBubble = NcCounterBubble.NcCounterBubble;
exports.NcDashboardWidget = NcDashboardWidget.NcDashboardWidget;
exports.NcDashboardWidgetItem = NcDashboardWidgetItem.NcDashboardWidgetItem;
exports.NcDateTime = NcDateTime.NcDateTime;
exports.NcDateTimePicker = Components_NcDateTimePicker;
exports.NcDateTimePickerNative = Components_NcDateTimePickerNative;
exports.NcDialog = NcDialog.NcDialog;
exports.NcDialogButton = NcDialogButton.NcDialogButton;
exports.NcEllipsisedOption = Components_NcEllipsisedOption;
exports.NcEmojiPicker = NcEmojiPicker.NcEmojiPicker;
exports.NcEmptyContent = Components_NcEmptyContent;
exports.NcFilePicker = NcFilePicker.NcFilePicker;
exports.NcFormBox = NcFormBox.NcFormBox;
exports.NcFormBoxButton = NcFormBoxButton.NcFormBoxButton;
exports.NcFormBoxCopyButton = NcFormBoxCopyButton.NcFormBoxCopyButton;
exports.NcFormBoxSwitch = NcFormBoxSwitch.NcFormBoxSwitch;
exports.NcFormGroup = NcFormGroup.NcFormGroup;
exports.NcGuestContent = Components_NcGuestContent;
exports.NcHeaderButton = NcHeaderButton.NcHeaderButton;
exports.NcHeaderMenu = NcHeaderMenu.NcHeaderMenu;
exports.NcHighlight = Components_NcHighlight.NcHighlight;
exports.NcHotkey = NcHotkey.NcHotkey;
exports.NcHotkeyList = NcHotkeyList.NcHotkeyList;
exports.NcIconSvgWrapper = NcIconSvgWrapper.NcIconSvgWrapper;
exports.NcInputField = NcInputField.NcInputField;
exports.NcListItem = NcListItem.NcListItem;
exports.NcListItemIcon = NcListItemIcon.NcListItemIcon;
exports.NcLoadingIcon = Components_NcLoadingIcon;
exports.NcModal = Components_NcModal;
exports.NcNoteCard = Components_NcNoteCard;
exports.NcPasswordField = NcPasswordField.NcPasswordField;
exports.NcPopover = NcPopover.NcPopover;
exports.NcProgressBar = NcProgressBar.NcProgressBar;
exports.NcRadioGroup = NcRadioGroup.NcRadioGroup;
exports.NcRadioGroupButton = NcRadioGroupButton.NcRadioGroupButton;
exports.NcRelatedResourcesPanel = NcRelatedResourcesPanel.NcRelatedResourcesPanel;
exports.NcMentionBubble = Mixins_richEditor.NcMentionBubble;
exports.richEditor = Mixins_richEditor.richEditor;
exports.NcAutoCompleteResult = NcRichContenteditable.NcAutoCompleteResult;
exports.NcRichContenteditable = NcRichContenteditable.NcRichContenteditable;
exports.anyLinkProviderId = referencePickerModal.anyLinkProviderId;
exports.getLinkWithPicker = referencePickerModal.getLinkWithPicker;
exports.getProvider = referencePickerModal.getProvider;
exports.getProviders = referencePickerModal.getProviders;
exports.searchProvider = referencePickerModal.searchProvider;
exports.sortProviders = referencePickerModal.sortProviders;
exports.NcCustomPickerRenderResult = customPickerElements.NcCustomPickerRenderResult;
exports.hasInteractiveView = customPickerElements.hasInteractiveView;
exports.isCustomPickerElementRegistered = customPickerElements.isCustomPickerElementRegistered;
exports.isWidgetRegistered = customPickerElements.isWidgetRegistered;
exports.registerCustomPickerElement = customPickerElements.registerCustomPickerElement;
exports.registerWidget = customPickerElements.registerWidget;
exports.renderCustomPickerElement = customPickerElements.renderCustomPickerElement;
exports.renderWidget = customPickerElements.renderWidget;
exports.NcRichText = NcRichText.NcRichText;
exports.NcSelect = NcSelect.NcSelect;
exports.NcTextField = NcTextField.NcTextField;
exports.NcSavingIndicatorIcon = NcSavingIndicatorIcon.NcSavingIndicatorIcon;
exports.NcSelectTags = NcSelectTags.NcSelectTags;
exports.NcSelectUsers = NcSelectUsers.NcSelectUsers;
exports.NcSettingsInputText = NcSettingsInputText.NcSettingsInputText;
exports.NcSettingsSection = NcSettingsSection.NcSettingsSection;
exports.NcSettingsSelectGroup = NcSettingsSelectGroup.NcSettingsSelectGroup;
exports.NcTextArea = NcTextArea.NcTextArea;
exports.NcThemeProvider = NcThemeProvider.NcThemeProvider;
exports.NcTimezonePicker = NcTimezonePicker.NcTimezonePicker;
exports.NcUserBubble = NcUserBubble.NcUserBubble;
exports.NcUserStatusIcon = NcUserStatusIcon.NcUserStatusIcon;
exports.NcVNodes = Components_NcVNodes;
exports.useFormatDateTime = Composables_useFormatDateTime.useFormatDateTime;
exports.useFormatRelativeTime = Composables_useFormatDateTime.useFormatRelativeTime;
exports.useFormatTime = Composables_useFormatDateTime.useFormatTime;
exports.useHotKey = Composables_useHotKey.useHotKey;
exports.useIsDarkTheme = Composables_useIsDarkTheme.useIsDarkTheme;
exports.useIsDarkThemeElement = Composables_useIsDarkTheme.useIsDarkThemeElement;
exports.isFullscreenState = Composables_useIsFullscreen.isFullscreenState;
exports.useIsFullscreen = Composables_useIsFullscreen.useIsFullscreen;
exports.MOBILE_BREAKPOINT = Composables_useIsMobile.MOBILE_BREAKPOINT;
exports.MOBILE_SMALL_BREAKPOINT = Composables_useIsMobile.MOBILE_SMALL_BREAKPOINT;
exports.isMobileState = Composables_useIsMobile.isMobileState;
exports.useIsMobile = Composables_useIsMobile.useIsMobile;
exports.useIsSmallMobile = Composables_useIsMobile.useIsSmallMobile;
exports.isA11yActivation = Functions_a11y.isA11yActivation;
exports.getEnabledContactsMenuActions = Functions_contactsMenu.getEnabledContactsMenuActions;
exports.registerContactsMenuAction = Functions_contactsMenu.registerContactsMenuAction;
exports.spawnDialog = Functions_dialog.spawnDialog;
exports.EmojiSkinTone = emoji.EmojiSkinTone;
exports.emojiAddRecent = emoji.emojiAddRecent;
exports.emojiSearch = emoji.emojiSearch;
exports.getCurrentSkinTone = emoji.getCurrentSkinTone;
exports.setCurrentSkinTone = emoji.setCurrentSkinTone;
exports.checkIfDarkTheme = Functions_isDarkTheme.checkIfDarkTheme;
exports.isDarkTheme = Functions_isDarkTheme.isDarkTheme;
exports.preloadImage = Functions_preloadImage.preloadImage;
exports.usernameToColor = Functions_usernameToColor.usernameToColor;
exports.Focus = Directives_Focus;
exports.Linkify = Directives_Linkify;
exports.clickOutsideOptions = Mixins_clickOutsideOptions;
exports.isFullscreen = Mixins_isFullscreen;
exports.isMobile = Mixins_isMobile;
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: () => floatingVue.VTooltip
});
exports.NextcloudVuePlugin = NextcloudVuePlugin;
