// @nextcloud/vue/dist/Functions/reference.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const customPickerElements = require("../chunks/customPickerElements-GyIIOHiQ.cjs");
const referencePickerModal = require("../chunks/referencePickerModal-BGYbZYrX.cjs");
exports.NcCustomPickerRenderResult = customPickerElements.NcCustomPickerRenderResult;
exports.hasInteractiveView = customPickerElements.hasInteractiveView;
exports.isCustomPickerElementRegistered = customPickerElements.isCustomPickerElementRegistered;
exports.isWidgetRegistered = customPickerElements.isWidgetRegistered;
exports.registerCustomPickerElement = customPickerElements.registerCustomPickerElement;
exports.registerWidget = customPickerElements.registerWidget;
exports.renderCustomPickerElement = customPickerElements.renderCustomPickerElement;
exports.renderWidget = customPickerElements.renderWidget;
exports.anyLinkProviderId = referencePickerModal.anyLinkProviderId;
exports.getLinkWithPicker = referencePickerModal.getLinkWithPicker;
exports.getProvider = referencePickerModal.getProvider;
exports.getProviders = referencePickerModal.getProviders;
exports.searchProvider = referencePickerModal.searchProvider;
exports.sortProviders = referencePickerModal.sortProviders;
