// @nextcloud/vue/dist/index.mjs
import { N } from "./chunks/NcActionButton-BJKJG2zO.mjs";
import { N as N2 } from "./chunks/NcActionButtonGroup-CgYSlRXJ.mjs";
import { N as N3 } from "./chunks/NcActionCaption-f7L9gBlT.mjs";
import { N as N4 } from "./chunks/NcActionCheckbox-D2Fu35M_.mjs";
import { N as N5 } from "./chunks/NcActionInput-gv6uLn_N.mjs";
import { N as N6 } from "./chunks/NcActionLink-BFiaYt9A.mjs";
import { N as N7 } from "./chunks/NcActionRadio-BJ4i9n8a.mjs";
import { N as N8 } from "./chunks/NcActionRouter-vYFtIOzD.mjs";
import { N as N9 } from "./chunks/NcActions-fCXSb_0o.mjs";
import { N as N10 } from "./chunks/NcActionSeparator-B9pNQaji.mjs";
import { N as N11 } from "./chunks/NcActionText-CQ9qwJ0p.mjs";
import { N as N12 } from "./chunks/NcActionTextEditable-Bd6fU8b3.mjs";
import { N as N13 } from "./chunks/NcAppContent-taTvqnvd.mjs";
import { _ } from "./chunks/NcAppContentDetails.vue_vue_type_script_setup_true_lang-BWl5Pw5g.mjs";
import { N as N14 } from "./chunks/NcAppContentList-Copx3Ver.mjs";
import { N as N15 } from "./chunks/NcAppNavigation-Bb54U_qJ.mjs";
import { N as N16 } from "./chunks/NcAppNavigationCaption-CwaRTmmd.mjs";
import { N as N17 } from "./chunks/NcAppNavigationIconBullet-BmU-dPcY.mjs";
import { N as N18 } from "./chunks/NcAppNavigationItem-Dg2d2AwS.mjs";
import { N as N19 } from "./chunks/NcAppNavigationList-CGSWabRB.mjs";
import { N as N20 } from "./chunks/NcAppNavigationNew-Ch_kPggd.mjs";
import { N as N21 } from "./chunks/NcAppNavigationNewItem-DYM8hEa7.mjs";
import { N as N22 } from "./chunks/NcAppNavigationSearch-BWHsAoc1.mjs";
import { N as N23 } from "./chunks/NcAppNavigationSettings-CJ2z5cYU.mjs";
import { N as N24 } from "./chunks/NcAppNavigationSpacer-CIRC4ZEj.mjs";
import { N as N25 } from "./chunks/NcAppSettingsDialog-37Ko8yBC.mjs";
import { N as N26 } from "./chunks/NcAppSettingsSection-MusCsTGg.mjs";
import { N as N27, N as N28 } from "./chunks/NcAppSettingsShortcutsSection-C0io5RJP.mjs";
import { N as N29 } from "./chunks/NcAppSidebar-CV2t1fDr.mjs";
import { _ as _2 } from "./chunks/NcAppSidebarHeader.vue_vue_type_script_setup_true_lang-Bk4yFBHY.mjs";
import { N as N30 } from "./chunks/NcAppSidebarTab-DOSDDbGA.mjs";
import { N as N31 } from "./chunks/NcAssistantButton-u8m6vfFY.mjs";
import { N as N32 } from "./chunks/NcAssistantContent-DWv2sak_.mjs";
import { N as N33 } from "./chunks/NcAssistantIcon-Cf0B3EsL.mjs";
import { N as N34 } from "./chunks/NcAvatar-QYahYt2p.mjs";
import { N as N35 } from "./chunks/NcBlurHash-B3MufoB_.mjs";
import { N as N36 } from "./chunks/NcBreadcrumb-CYI3eykz.mjs";
import { N as N37 } from "./chunks/NcBreadcrumbs-DLKGhmX6.mjs";
import { N as N38 } from "./chunks/NcButton-CdQzGSB6.mjs";
import { N as N39 } from "./chunks/NcCheckboxRadioSwitch-B0KUiKrb.mjs";
import { N as N40 } from "./chunks/NcChip-CjYZUIUu.mjs";
import { N as N41 } from "./chunks/NcCollectionList-CU1Tigay.mjs";
import { N as N42 } from "./chunks/NcColorPicker-DUIppkpO.mjs";
import { N as N43 } from "./chunks/NcContent-Dd15hgck.mjs";
import { N as N44 } from "./chunks/NcCounterBubble-VUNXKsnk.mjs";
import { N as N45 } from "./chunks/NcDashboardWidget-CPI2VARK.mjs";
import { N as N46 } from "./chunks/NcDashboardWidgetItem-DJ820dYQ.mjs";
import { _ as _3 } from "./chunks/NcDateTime.vue_vue_type_script_setup_true_lang-B4upiZjL.mjs";
import { N as N47 } from "./chunks/NcDateTimePicker-8vIvIKoD.mjs";
import { N as N48 } from "./chunks/NcDateTimePickerNative-XO_O44n2.mjs";
import { N as N49 } from "./chunks/NcDialog-D-qxVQxL.mjs";
import { _ as _4 } from "./chunks/NcDialogButton.vue_vue_type_script_setup_true_lang-WZfDqDwc.mjs";
import { N as N50 } from "./chunks/NcEllipsisedOption-D6Amb91K.mjs";
import { N as N51 } from "./chunks/NcEmojiPicker-DG3qO_fi.mjs";
import { N as N52 } from "./chunks/NcEmptyContent--TRhmNAW.mjs";
import { N as N53 } from "./chunks/NcFilePicker-DPiYJu5B.mjs";
import { N as N54 } from "./chunks/NcFormBox-wP5mpW88.mjs";
import { N as N55 } from "./chunks/NcFormBoxButton-DhEpFn7z.mjs";
import { _ as _5 } from "./chunks/NcFormBoxCopyButton.vue_vue_type_script_setup_true_lang-DLh3Ylq_.mjs";
import { N as N56 } from "./chunks/NcFormBoxSwitch-QMisnf8C.mjs";
import { N as N57 } from "./chunks/NcFormGroup-BDjfDvq0.mjs";
import { N as N58 } from "./chunks/NcGuestContent-_z22nZMF.mjs";
import { N as N59 } from "./chunks/NcHeaderButton-lhDrtYWN.mjs";
import { N as N60 } from "./chunks/NcHeaderMenu-B5-BQbQ6.mjs";
import { _ as _6 } from "./chunks/NcHighlight.vue_vue_type_script_lang-DnWQDM_2.mjs";
import { N as N61 } from "./chunks/NcHotkey-B0qDRy1t.mjs";
import { N as N62 } from "./chunks/NcHotkeyList-Ca7AtgZ9.mjs";
import { N as N63 } from "./chunks/NcIconSvgWrapper-De-2-ukl.mjs";
import { N as N64 } from "./chunks/NcInputField-C3iof9pY.mjs";
import { N as N65 } from "./chunks/NcKbd-Caeq706M.mjs";
import { N as N66 } from "./chunks/NcListItem-Chk9TtPa.mjs";
import { N as N67 } from "./chunks/NcListItemIcon-vZ-OTxIa.mjs";
import { N as N68 } from "./chunks/NcLoadingIcon-CInLzPtA.mjs";
import { N as N69 } from "./chunks/NcModal-DyOZxq9E.mjs";
import { N as N70 } from "./chunks/NcNoteCard-BJF9bD7S.mjs";
import { N as N71 } from "./chunks/NcPasswordField-ykWHv_hw.mjs";
import { N as N72 } from "./chunks/NcPopover-CtdLAkEU.mjs";
import { N as N73 } from "./chunks/NcProgressBar-JyXXoHLT.mjs";
import { N as N74 } from "./chunks/NcRadioGroup-EGc_x6uH.mjs";
import { N as N75 } from "./chunks/NcRadioGroupButton-Dd5hMWx5.mjs";
import { N as N76 } from "./chunks/NcRelatedResourcesPanel-C7C0-pBV.mjs";
import { N as N77, a, b } from "./chunks/NcRichContenteditable-DLPNb2TV.mjs";
import "@nextcloud/auth";
import "@nextcloud/axios";
import "@nextcloud/router";
import "@nextcloud/sharing/public";
import "vue";
import "@vueuse/core";
import "vue-router";
import { a as a2, g, b as b2, c, s, d } from "./chunks/referencePickerModal-DgIpbEdl.mjs";
import { N as N78, h, i, b as b3, r, c as c2, e, f } from "./chunks/customPickerElements-Cu7bLbap.mjs";
import "unist-builder";
import "unist-util-visit-parents";
import "./chunks/logger-D3RVzcfQ.mjs";
import { N as N79 } from "./chunks/NcRichText-rS9gcduo.mjs";
import { N as N80 } from "./chunks/NcSelect-DK4ZWMha.mjs";
import "debounce";
import { _ as _7 } from "./chunks/NcTextField.vue_vue_type_script_setup_true_lang-DgvCZDmR.mjs";
import "@nextcloud/event-bus";
import { _ as _8 } from "./chunks/NcSavingIndicatorIcon.vue_vue_type_script_setup_true_lang-CA0r_04W.mjs";
import { N as N81 } from "./chunks/NcSelectTags-D3tAMWYx.mjs";
import { N as N82 } from "./chunks/NcSelectUsers-2L4I2yvb.mjs";
import { N as N83 } from "./chunks/NcSettingsSection-DnzR8DRc.mjs";
import { N as N84 } from "./chunks/NcSettingsSelectGroup-CYgEv80h.mjs";
import { N as N85 } from "./chunks/NcTextArea-BwFQx9Bj.mjs";
import { _ as _9 } from "./chunks/NcThemeProvider.vue_vue_type_script_setup_true_lang-wie70gn4.mjs";
import { _ as _10 } from "./chunks/NcTimezonePicker.vue_vue_type_script_setup_true_lang-53QSdjMX.mjs";
import { N as N86 } from "./chunks/NcUserBubble-BnaGHDoD.mjs";
import { N as N87 } from "./chunks/NcUserStatusIcon-CNh1vXUF.mjs";
import { _ as _11 } from "./chunks/NcVNodes.vue_vue_type_script_lang-BqUHinRZ.mjs";
import { useFormatRelativeTime, useFormatTime } from "./composables/useFormatDateTime/index.mjs";
import { useHotKey } from "./composables/useHotKey/index.mjs";
import { useIsDarkTheme, useIsDarkThemeElement } from "./composables/useIsDarkTheme/index.mjs";
import { useIsFullscreen } from "./composables/useIsFullscreen/index.mjs";
import { MOBILE_BREAKPOINT, MOBILE_SMALL_BREAKPOINT, useIsMobile, useIsSmallMobile } from "./composables/useIsMobile/index.mjs";
import { isA11yActivation } from "./functions/a11y/index.mjs";
import { getEnabledContactsMenuActions, registerContactsMenuAction } from "./functions/contactsMenu/index.mjs";
import { spawnDialog } from "./functions/dialog/index.mjs";
import { E, e as e2, a as a3, g as g2, s as s2 } from "./chunks/emoji-C8k9NUlo.mjs";
import { checkIfDarkTheme, isDarkTheme } from "./functions/isDarkTheme/index.mjs";
import { preloadImage } from "./functions/preloadImage/index.mjs";
import { usernameToColor } from "./functions/usernameToColor/index.mjs";
import { default as default2 } from "./directives/Focus/index.mjs";
import { default as default3 } from "./directives/Linkify/index.mjs";
export {
  E as EmojiSkinTone,
  default2 as Focus,
  default3 as Linkify,
  MOBILE_BREAKPOINT,
  MOBILE_SMALL_BREAKPOINT,
  N as NcActionButton,
  N2 as NcActionButtonGroup,
  N3 as NcActionCaption,
  N4 as NcActionCheckbox,
  N5 as NcActionInput,
  N6 as NcActionLink,
  N7 as NcActionRadio,
  N8 as NcActionRouter,
  N10 as NcActionSeparator,
  N11 as NcActionText,
  N12 as NcActionTextEditable,
  N9 as NcActions,
  N13 as NcAppContent,
  _ as NcAppContentDetails,
  N14 as NcAppContentList,
  N15 as NcAppNavigation,
  N16 as NcAppNavigationCaption,
  N17 as NcAppNavigationIconBullet,
  N18 as NcAppNavigationItem,
  N19 as NcAppNavigationList,
  N20 as NcAppNavigationNew,
  N21 as NcAppNavigationNewItem,
  N22 as NcAppNavigationSearch,
  N23 as NcAppNavigationSettings,
  N24 as NcAppNavigationSpacer,
  N25 as NcAppSettingsDialog,
  N26 as NcAppSettingsSection,
  N27 as NcAppSettingsSectionShortcuts,
  N28 as NcAppSettingsShortcutsSection,
  N29 as NcAppSidebar,
  _2 as NcAppSidebarHeader,
  N30 as NcAppSidebarTab,
  N31 as NcAssistantButton,
  N32 as NcAssistantContent,
  N33 as NcAssistantIcon,
  N77 as NcAutoCompleteResult,
  N34 as NcAvatar,
  N35 as NcBlurHash,
  N36 as NcBreadcrumb,
  N37 as NcBreadcrumbs,
  N38 as NcButton,
  N39 as NcCheckboxRadioSwitch,
  N40 as NcChip,
  N41 as NcCollectionList,
  N42 as NcColorPicker,
  N43 as NcContent,
  N44 as NcCounterBubble,
  N78 as NcCustomPickerRenderResult,
  N45 as NcDashboardWidget,
  N46 as NcDashboardWidgetItem,
  _3 as NcDateTime,
  N47 as NcDateTimePicker,
  N48 as NcDateTimePickerNative,
  N49 as NcDialog,
  _4 as NcDialogButton,
  N50 as NcEllipsisedOption,
  N51 as NcEmojiPicker,
  N52 as NcEmptyContent,
  N53 as NcFilePicker,
  N54 as NcFormBox,
  N55 as NcFormBoxButton,
  _5 as NcFormBoxCopyButton,
  N56 as NcFormBoxSwitch,
  N57 as NcFormGroup,
  N58 as NcGuestContent,
  N59 as NcHeaderButton,
  N60 as NcHeaderMenu,
  _6 as NcHighlight,
  N61 as NcHotkey,
  N62 as NcHotkeyList,
  N63 as NcIconSvgWrapper,
  N64 as NcInputField,
  N65 as NcKbd,
  N66 as NcListItem,
  N67 as NcListItemIcon,
  N68 as NcLoadingIcon,
  a as NcMentionBubble,
  N69 as NcModal,
  N70 as NcNoteCard,
  N71 as NcPasswordField,
  N72 as NcPopover,
  N73 as NcProgressBar,
  N74 as NcRadioGroup,
  N75 as NcRadioGroupButton,
  N76 as NcRelatedResourcesPanel,
  b as NcRichContenteditable,
  N79 as NcRichText,
  _8 as NcSavingIndicatorIcon,
  N80 as NcSelect,
  N81 as NcSelectTags,
  N82 as NcSelectUsers,
  N83 as NcSettingsSection,
  N84 as NcSettingsSelectGroup,
  N85 as NcTextArea,
  _7 as NcTextField,
  _9 as NcThemeProvider,
  _10 as NcTimezonePicker,
  N86 as NcUserBubble,
  N87 as NcUserStatusIcon,
  _11 as NcVNodes,
  a2 as anyLinkProviderId,
  checkIfDarkTheme,
  e2 as emojiAddRecent,
  a3 as emojiSearch,
  g2 as getCurrentSkinTone,
  getEnabledContactsMenuActions,
  g as getLinkWithPicker,
  b2 as getProvider,
  c as getProviders,
  h as hasInteractiveView,
  isA11yActivation,
  i as isCustomPickerElementRegistered,
  isDarkTheme,
  b3 as isWidgetRegistered,
  preloadImage,
  registerContactsMenuAction,
  r as registerCustomPickerElement,
  c2 as registerWidget,
  e as renderCustomPickerElement,
  f as renderWidget,
  s as searchProvider,
  s2 as setCurrentSkinTone,
  d as sortProviders,
  spawnDialog,
  useFormatRelativeTime,
  useFormatTime,
  useHotKey,
  useIsDarkTheme,
  useIsDarkThemeElement,
  useIsFullscreen,
  useIsMobile,
  useIsSmallMobile,
  usernameToColor
};
