// @nextcloud/vue/dist/Components/NcRichText.cjs
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
const NcRichText = require("../chunks/NcRichText-pqvg5l27.cjs");
const referencePickerModal = require("../chunks/referencePickerModal-BGYbZYrX.cjs");
const customPickerElements = require("../chunks/customPickerElements-GyIIOHiQ.cjs");
exports.NcReferenceList = NcRichText.NcReferenceList;
exports.NcRichText = NcRichText.NcRichText;
exports.default = NcRichText.NcRichText;
exports.NcReferencePicker = referencePickerModal.NcReferencePicker;
exports.NcReferencePickerModal = referencePickerModal.NcReferencePickerModal;
exports.NcReferenceWidget = referencePickerModal.NcReferenceWidget;
exports.NcSearch = referencePickerModal.NcSearch;
exports.anyLinkProviderId = referencePickerModal.anyLinkProviderId;
exports.getLinkWithPicker = referencePickerModal.getLinkWithPicker;
exports.getProvider = referencePickerModal.getProvider;
exports.getProviders = referencePickerModal.getProviders;
exports.searchProvider = referencePickerModal.searchProvider;
exports.sortProviders = referencePickerModal.sortProviders;
exports.NcCustomPickerRenderResult = customPickerElements.NcCustomPickerRenderResult;
exports.isCustomPickerElementRegistered = customPickerElements.isCustomPickerElementRegistered;
exports.isWidgetRegistered = customPickerElements.isWidgetRegistered;
exports.registerCustomPickerElement = customPickerElements.registerCustomPickerElement;
exports.registerWidget = customPickerElements.registerWidget;
exports.renderCustomPickerElement = customPickerElements.renderCustomPickerElement;
exports.renderWidget = customPickerElements.renderWidget;
