// @ckpack/vue-color/libs/index.js
export { default, default as install } from './preset.js';
export { default as create } from './create.js';
export { default as ColorMixin } from './mixin/color.js';
export { components } from './components.js';
export { default as Alpha } from './components/alpha/index.js';
export { default as Checkboard } from './components/checkboard/index.js';
export { default as Chrome } from './components/chrome/index.js';
export { default as Compact } from './components/compact/index.js';
export { default as EditableInput } from './components/editable-input/index.js';
export { default as Grayscale } from './components/grayscale/index.js';
export { default as Hue } from './components/hue/index.js';
export { default as Material } from './components/material/index.js';
export { default as Photoshop } from './components/photoshop/index.js';
export { default as Saturation } from './components/saturation/index.js';
export { default as Sketch } from './components/sketch/index.js';
export { default as Slider } from './components/slider/index.js';
export { default as Swatches } from './components/swatches/index.js';
export { default as Twitter } from './components/twitter/index.js';
import './defaultConfig.js';
import '@ctrl/tinycolor';
import 'vue';
import './style-inject.es-746bb8ed.js';
import './utils/compoent.js';
import './utils/utils.js';
import 'material-colors';
